### PR TITLE
Feature/admission contract

### PR DESCRIPTION
**Pull request by NikoBergemann**
### Reason for this PR
- We need to support Admissions on Blockchain

### Changes in this PR
- add Admission Model
- add Admission Contract + ContractUtil
- add Admission Tests
- refactor Test Creation in base class
- code cleanup
- remove duplicate code
- update github workflows to always upload coverage report, even on failure

### CODE REVIEW

**Review by anemonekk — APPROVED** [review]

**Review by mco33415 — APPROVED** [review]

File: UC4-chaincode/src/test/java/de/upb/cs/uc4/chaincode/TestCreationBase.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import java.io.FileReader;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class TestCreationBase {
+
+    abstract DynamicTest CreateTest(JsonIOTest test);
+
+    abstract String GetTestConfigDir();
```
Comment (matthias-geuchen) [inline UC4-chaincode/src/test/java/de/upb/cs/uc4/chaincode/TestCreationBase.java:21]:
```suggestion
    abstract String getTestConfigDir();
```
This is the java way, i think...

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/AdmissionContract.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        ArrayList<InvalidParameter> invalidParams = cUtil.getParameterErrorsForAdmission(newAdmission);
+        if (!invalidParams.isEmpty()) {
+            return GsonWrapper.toJson(cUtil.getUnprocessableEntityError(invalidParams));
+        }
+
+        if (cUtil.keyExists(stub, newAdmission.getAdmissionId())) {
+            return GsonWrapper.toJson(cUtil.getConflictError());
+        }
```
Comment (matthias-geuchen) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/AdmissionContract.java:55]: Is there a reason to check for conflict twice?

**Review by matthias-geuchen — CHANGES_REQUESTED** [review]

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/AdmissionContract.java (hunk 2 above)
Comment (NikoBergemann) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/AdmissionContract.java:55]:
```suggestion
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/ContractUtil.java [hunk 3]
```diff
@@ -1,39 +1,40 @@
...
-    }
-
-    public GenericError getConflictError() {
-        return null;
+    public DetailedError getInvalidActionError(ArrayList<InvalidParameter> invalidParams) {
+        return new DetailedError()
+                .type("HLInvalidEntity")
+                .title("The following parameters produced some semantic error")
+                .invalidParams(invalidParams);
     }
```
[the remark this answers is not in the document]
Comment (NikoBergemann) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/ContractUtil.java:35]:
```suggestion
    public DetailedError getInvalidActionError(ArrayList<InvalidParameter> invalidParams) {
        return new DetailedError()
                .type("HLUnprocessibleEntity")
                .title("The following parameters produced some semantic error")
                .invalidParams(invalidParams);
    }
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/ApprovalContractUtil.java [hunk 4]
```diff
@@ -33,24 +33,17 @@ public GenericError getNotFoundError() {
         return super.getNotFoundError(thing, identifier);
     }
 
+    @Override
+    public GenericError getConflictError() {
+        return super.getConflictError(thing, identifier);
+    }
```
[the remark this answers is not in the document]
Comment (NikoBergemann) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/ApprovalContractUtil.java]: all outsourced to general contract util, parameterized specialContractUtil

**Pull request merged**